### ISSUE TITLE
[no-unsafe-assignment] is is possible to create flag to disable this rule for `catch` blocks?

### ISSUE THREAD

**Issue by gerardolima** [typescript-eslint/typescript-eslint#3348]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": ["error"]
  }
}
```

```TS
try {
  throw new Error("any-error")
}
catch(err) {
  console.log({err}) // lint error (why?)
}
```

**Expected Result**
The behaviour is consistent with documentation.
I hoped there could exist a way _NOT_ have one error here.

**Actual Result**
Lint error (as documented).

**Additional Info**
The lack of an exception for `catch` blocks firstly led me to creating unnecessary variables such as `typedErr` (below). This boilerplate eventually annoyed me enough and I turned this rule off (which I didn't like to). So, I wonder if it would be of interest of @typescript-eslint to create an exception condition for this rule. 

```
try {
  throw new Error("any-error")
}
catch(err) {
  const typedErr = err as Error
  console.log({typedErr})
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^4.16.1` |
| `@typescript-eslint/parser`        | `^4.16.1` |
| `TypeScript`                       | `^4.2.3` |
| `ESLint`                           | `^7.21.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
the variable is of type `any` - assigning it to anything is unsafe.

In JS there's no requirement for a caught exception to be of any specific type.
For example:
```ts
throw 'a';
throw 1;
throw false;
throw () => {};
throw new Promise();
```
Are all 100% valid in JS.

An assertion is inherently unsafe as it's not checked by the compiler at all.
Sometimes you can reasonably infer the type of what an exception will likely be, though you can never be 100% certain when you're calling code across module boundaries - or even across function boundaries in the same file.

If you don't want to be working with `any`, then just annotate your caught error with `unknown`, which is supported in TS:

```ts
try {
  throw new Error("any-error")
}
catch(err: unknown) {
  console.log({err}) // no lint error
}
```

From `unknown` you can reasonably refine the type and keep your code safe:
```ts
try {
  throw new Error("any-error")
}
catch(err: unknown) {
  if (err instanceof Error) {
    // err is guaranteed at type-time and runtime to only ever be an Error object
  } else {
    // err is of some unknown form that we don't know how to handle
    throw err;
  }
}
```

**Comment by gerardolima** [comment]
hey, @bradzacher,

I understand the reasoning on the rule, but still I think that `catch(err)` blocks should infer type of err as `unknown` (which it _is_), instead of `any`, whose usage is highly discouraged.

As this might not correspond to the type inference on tsc, I suggest this behavior to be made available using a flag, as shown below:

```json
{
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": ["error", { "inferCatchAsUnknown": true } ]
  }
}
```

**Comment by bradzacher** [comment]
> but still I think that catch(err) blocks should infer type of err as unknown 

I agree. But the typescript authors have chosen not to do it by default and not to do it via a flag.

Instead in v4.0 they added the ability to annotate your catch causes as unknown:

```ts
try {}
catch (e: unknown) {}
```

We even have the rule [no-implicit-any-catch](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-implicit-any-catch.md) to help you remember to annotate your catch variables.

**Comment by gerardolima** [comment]
I think the rule  no-implicit-any-catch might be the workaround that will solve my issue. Thank you, @bradzacher!!
